### PR TITLE
Allow passing an array as intensity for plot3d

### PR DESCRIPTION
**Pull request by core-man**
**Description of proposed changes**

Currently, we can have two ways to specify an intensity/intensities using the [plot3d](https://www.pygmt.org/latest/api/generated/pygmt.Figure.plot3d.html) method.
- Simply give a constant intensity like `intensity=0.5`, which works for both `x`/`y`/`z` and `data` input. This is similar to `-I0.5` in the GMT command line.
- Set `intensity=True`, and read intensities from an *input file* or from a *2D array*, which is only for the `data` parameter. This is similar to `-I` in the GMT command line.

So, when we want to change the intensities for different points, we can only use the `data` parameter now. This PR tries to pass an array to `intensity` so that we can set different intensities when we use `x`/`y`/`z` as point input. This way will not work for `data` input.

Related to #891 

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/src/plot3d.py [hunk 1]
```diff
@@ -111,11 +111,14 @@ def plot3d(
...
-        modulate the fill color by simulating illumination [Default is None].
-        If using ``intensity=True``, we will instead read *intens* from the
-        first data column after the symbol parameters (if given).
+    intensity : float or bool or 1d array
+        Provide an *intensity* value (nominally in the -1 to +1 range) to
+        modulate the fill color by simulating illumination. If using
+        ``intensity=True``, we will instead read *intensity* from the first
+        data column after the symbol parameters (if given). *intensity* can
+        also be a 1d array to set varying intensity for symbols, but it is only
+        valid for ``x``/``y`` pairs.
```
Comment (seisman) [inline pygmt/src/plot3d.py:120]:
```suggestion
        valid for ``x``/``y``/``z``.
```

File: pygmt/tests/test_plot3d.py [hunk 2]
```diff
@@ -329,6 +318,34 @@ def test_plot3d_colors_sizes_proj(data, region):
...
+    z = y
+    intensity = x
+
+    fig = Figure()
+    fig.plot3d(
+        x=x,
+        y=y,
+        z=z,
+        region=[-1.1, 1.1, -0.5, 0.5, -0.5, 0.5],
+        projection="x15c",
```
Comment (seisman) [inline pygmt/tests/test_plot3d.py:338]: Any particular reason that you want to use `x15c` (not `X15c`) here?
```suggestion
        projection="x15c",
```

Comment (core-man) [reply pygmt/tests/test_plot3d.py]: Revised in 3c12707.

File: pygmt/tests/test_plot3d.py [hunk 3]
```diff
@@ -329,6 +318,34 @@ def test_plot3d_colors_sizes_proj(data, region):
...
+    z = y
+    intensity = x
+
+    fig = Figure()
+    fig.plot3d(
+        x=x,
+        y=y,
+        z=z,
+        region=[-1.1, 1.1, -0.5, 0.5, -0.5, 0.5],
+        projection="X15c",
```
Comment (seisman) [inline pygmt/tests/test_plot3d.py:338]:
```suggestion
        projection="X15c/5c",
```

File: pygmt/tests/test_plot3d.py [hunk 4]
```diff
@@ -329,6 +318,34 @@ def test_plot3d_colors_sizes_proj(data, region):
...
+        zscale=0.1,
```
Comment (seisman) [inline pygmt/tests/test_plot3d.py:339]:
```suggestion
        zsize="5c",
```

**Review by seisman — APPROVED** [review]

File: pygmt/tests/test_plot3d.py [hunk 5]
```diff
@@ -329,6 +318,34 @@ def test_plot3d_colors_sizes_proj(data, region):
     return fig_ref, fig_test
 
 
+@pytest.mark.mpl_image_compare
+def test_plot3d_varying_intensity():
+    """
+    Plot the data with array-like intensity.
+    """
+
```
Comment (seisman) [inline pygmt/tests/test_plot3d.py:326]:
```suggestion
    """
```

File: pygmt/tests/test_plot3d.py [hunk 6]
```diff
@@ -329,6 +318,34 @@ def test_plot3d_colors_sizes_proj(data, region):
...
+    x = np.arange(-1, 1.1, 0.1)
```
Comment (seisman) [inline pygmt/tests/test_plot3d.py:327]:
```suggestion
    """
    x = np.arange(-1, 1.1, 0.1)
```

**Pull request merged**